File: test/vault_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:ohmyssh/data/models.dart';
import 'package:ohmyssh/data/vault.dart';

void main() {
  late Directory temp;

  setUp(() async {
    temp = await Directory.systemTemp.createTemp('ohmyssh_vault_test');
  });

  tearDown(() async {
    if (await temp.exists()) await temp.delete(recursive: true);
  });

  File vaultFile() => File('${temp.path}/test.vault');

  VaultData sample() => VaultData(
    hosts: [
      const Host(
        id: 'h1',
        label: 'web-01',
        hostname: '10.0.0.5',
        port: 2222,
        identityId: 'i1',
        knownHostKey: 'SHA256:abc',
        osId: 'ubuntu',
      ),
    ],
    identities: [
      const Identity(
        id: 'i1',
        label: 'root',
        username: 'root',
        kind: AuthKind.password,
        password: 'hunter2',
      ),
    ],
    groups: [const HostGroup(id: 'g1', name: 'Production')],
  );

  test('round-trips data through create + open', () async {
    final vault = await Vault.create(
      password: 'correct horse battery',
      file: vaultFile(),
      initial: sample(),
    );
    expect(await vaultFile().exists(), isTrue);

    final reopened = await Vault.open(
      password: 'correct horse battery',
      file: vaultFile(),
    );
    final data = await reopened.read();

    expect(data.hosts, hasLength(1));
    expect(data.hosts.single.hostname, '10.0.0.5');
    expect(data.hosts.single.port, 2222);
    expect(data.hosts.single.knownHostKey, 'SHA256:abc');
    expect(data.identities.single.password, 'hunter2');
    expect(data.groups.single.name, 'Production');

    await vault.save(data);
  });

  test('rejects the wrong password', () async {
    await Vault.create(
      password: 'right',
      file: vaultFile(),
      initial: sample(),
    );

    expect(
      () => Vault.open(password: 'wrong', file: vaultFile()),
      throwsA(isA<WrongPasswordException>()),
    );
  });

  test('secrets are not readable in the file on disk', () async {
    await Vault.create(
      password: 'right',
      file: vaultFile(),
      initial: sample(),
    );

    final raw = await vaultFile().readAsString();
    expect(raw, isNot(contains('hunter2')));
    expect(raw, isNot(contains('10.0.0.5')));
    expect(raw, isNot(contains('web-01')));
    expect(raw, contains(kVaultFormat));
  });

  test('changePassword re-encrypts without losing data', () async {
    final vault = await Vault.create(
      password: 'old-password',
      file: vaultFile(),
      initial: sample(),
    );

    await vault.changePassword('new-password');

    expect(
      () => Vault.open(password: 'old-password', file: vaultFile()),
      throwsA(isA<WrongPasswordException>()),
    );

    final reopened = await Vault.open(
      password: 'new-password',
      file: vaultFile(),
    );
    expect((await reopened.read()).hosts.single.label, 'web-01');
  });

  test('refuses a file that is not an ohmyssh vault', () async {
    await vaultFile().writeAsString('{"format":"something-else"}');
    expect(
      () => Vault.open(password: 'x', file: vaultFile()),
      throwsA(isA<VaultException>()),
    );
  });
}

File: shared/src/desktopTest/kotlin/com/example/ohmyssh/VaultTest.kt
package com.example.ohmyssh

import com.example.ohmyssh.data.AuthKind
import com.example.ohmyssh.data.Host
import com.example.ohmyssh.data.HostGroup
import com.example.ohmyssh.data.Identity
import com.example.ohmyssh.data.Vault
import com.example.ohmyssh.data.VaultData
import com.example.ohmyssh.data.VaultException
import com.example.ohmyssh.data.WrongPasswordException
import com.example.ohmyssh.data.kVaultFormat
import kotlinx.coroutines.runBlocking
import java.io.File
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class VaultTest {
    private lateinit var temp: File

    @BeforeTest
    fun setUp() {
        temp = File.createTempFile("ohmyssh_vault_test", "").apply {
            delete()
            mkdirs()
        }
    }

    @AfterTest
    fun tearDown() {
        temp.deleteRecursively()
    }

    private fun vaultPath(): String = File(temp, "test.vault").path

    private fun sample() = VaultData(
        hosts = listOf(
            Host(
                id = "h1",
                label = "web-01",
                hostname = "10.0.0.5",
                port = 2222,
                identityId = "i1",
                knownHostKey = "SHA256:abc",
                osId = "ubuntu",
            ),
        ),
        identities = listOf(
            Identity(
                id = "i1",
                label = "root",
                username = "root",
                kind = AuthKind.PASSWORD,
                password = "hunter2",
            ),
        ),
        groups = listOf(HostGroup(id = "g1", name = "Production")),
    )

    @Test
    fun roundTripsDataThroughCreateAndOpen() = runBlocking {
        val path = vaultPath()
        val vault = Vault.create("correct horse battery", path, sample())
        assertTrue(File(path).exists())

        val reopened = Vault.open("correct horse battery", path)
        val data = reopened.read()

        assertEquals(1, data.hosts.size)
        assertEquals("10.0.0.5", data.hosts.single().hostname)
        assertEquals(2222, data.hosts.single().port)
        assertEquals("SHA256:abc", data.hosts.single().knownHostKey)
        assertEquals("hunter2", data.identities.single().password)
        assertEquals("Production", data.groups.single().name)

        vault.save(data)
    }

    @Test
    fun rejectsTheWrongPassword() = runBlocking {
        val path = vaultPath()
        Vault.create("right", path, sample())

        assertFailsWith<WrongPasswordException> { Vault.open("wrong", path) }
        Unit
    }

    @Test
    fun secretsAreNotReadableInTheFileOnDisk() = runBlocking {
        val path = vaultPath()
        Vault.create("right", path, sample())

        val raw = File(path).readText()
        assertFalse(raw.contains("hunter2"))
        assertFalse(raw.contains("10.0.0.5"))
        assertFalse(raw.contains("web-01"))
        assertContains(raw, kVaultFormat)
    }

    @Test
    fun changePasswordReEncryptsWithoutLosingData() = runBlocking {
        val path = vaultPath()
        val vault = Vault.create("old-password", path, sample())

        vault.changePassword("new-password")

        assertFailsWith<WrongPasswordException> { Vault.open("old-password", path) }

        val reopened = Vault.open("new-password", path)
        assertEquals("web-01", reopened.read().hosts.single().label)
    }

    @Test
    fun refusesAFileThatIsNotAnOhmysshVault() = runBlocking {
        val path = vaultPath()
        File(path).writeText("""{"format":"something-else"}""")
        assertFailsWith<VaultException> { Vault.open("x", path) }
        Unit
    }
}
